### PR TITLE
Introduce ModelManagerThrowable

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

There is an extension point to handle differently the ModelManagerException in the CRUDController: 
https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/Controller/CRUDController.php#L1006

A nice way to have different behaviour was to extend the ModelManagerException and doing some instanceof checks.
ModelManagerException is final in 4.x. So introducing the interface, allow to implements it instead.

I didn't introduce interfaces for others exception because I didn't really find the needs. But same could be done

## Changelog

```markdown
### Added
- `ModelManagerThrowable`
- `CRUDController::handleModelManagerThrowable()`

### Deprecated
- `CRUDController::handleModelManagerException()`
```

### CODE REVIEW

**Comment by VincentLanglet** [comment]
Changing the signature of 
```
protected function handleModelManagerException(\Exception $e)
```
would be a BC-break.

What are my solutions ?

**Comment by core23** [comment]
> A nice way to have different behaviour was to extend the ModelManagerException and doing some instanceof checks.
> ModelManagerException is final in 4.x. So introducing the interface, allow to implements it instead.

Why not making the class non-final and change the class to an interface in 5.0?

**Comment by VincentLanglet** [comment]
> Why not making the class non-final and change the class to an interface in 5.0?

That was my last solution, I wanted to try first to find a solution with a BC-layer/some deprecations and I think i found a way.

I added a commit that might work.
The idea is to promote using `handleModelManagerThrowable` over `handleModelManagerException`.

**Comment by core23** [comment]
> That was my last solution, I wanted to try first to find a solution with a BC-layer/some deprecations and I think i found a way.

What about moving the exception next to the manager?

- `Sonata\AdminBundle\Model\ModelManagerException` instead of `Sonata\AdminBundle\Exception\ModelManagerException`

This way we can introduce an interface without breaking BC.

**Comment by VincentLanglet** [comment]
> What about moving the exception next to the manager?
> 
> * `Sonata\AdminBundle\Model\ModelManagerException` instead of `Sonata\AdminBundle\Exception\ModelManagerException`

Not sure about it
Currently all the exception are in the the Exception folder, it could be weird to change this.

> This way we can introduce an interface without breaking BC.

I don't understand how it avoid BC break

**Comment by core23** [comment]
> Not sure about it
> Currently all the exception are in the the Exception folder, it could be weird to change this.

That's not completely right:
https://github.com/sonata-project/SonataAdminBundle/blob/4.x/src/Form/Extension/ChoiceTypeExtension.php

> > This way we can introduce an interface without breaking BC.
> 
> I don't understand how it avoid BC break

Instead of creating the `ModelManagerThrowable` interface, you can just create a `ModelManagerException` interface in an other namespace. This way we can keep the `Exception` suffix instead of introducing a new `Throwable` suffix.

**Comment by VincentLanglet** [comment]
> > > This way we can introduce an interface without breaking BC.
> > 
> > 
> > I don't understand how it avoid BC break
> 
> Instead of creating the `ModelManagerThrowable` interface, you can just create a `ModelManagerException` interface in an other namespace. This way we can keep the `Exception` suffix instead of introducing a new `Throwable` suffix.

The issue is not about the `handleModelManagerException` name of the method. But the signature.
It currently only accept `\Exception` as param and not `\Throwable`.

**Review by jordisala1991 — DISMISSED** [review]

File: src/Controller/CRUDController.php
Comment (phansys) [inline src/Controller/CRUDController.php]:
```suggestion
            'The method "%s()" is deprecated since sonata-project/admin-bundle 3.107 and will be removed in 5.0.',
```

Comment (phansys) [inline src/Controller/CRUDController.php]: IMO, we must keep the responsibility for these changes to the pre-release PRs.
```suggestion
            'The method %s is deprecated since sonata-project/admin-bundle 3.x and will be removed in 5.0.',
```

**Review by phansys — DISMISSED** [review]

File: src/Controller/CRUDController.php
Comment (VincentLanglet) [reply src/Controller/CRUDController.php]: > IMO, we must keep the responsibility for these changes to the pre-release PRs.

What's the reason behind ? In that particular case, I wanted to gain some time

Comment (phansys) [reply src/Controller/CRUDController.php]: I understand your motivation.
I think the reason behind this rule is to keep the changes agnostic to a specific version, as we don't know which release will include them until the PR is merged. By instance, we could decide to release 3.107 before and if this rule is honored, the PR should be still compatible with any future version.

Comment (VincentLanglet) [reply src/Controller/CRUDController.php]: Sure, but I think we already merged PR with number of version when the number of version was sure, it's not like there was any other PR for new feature on 3.x

Comment (VincentLanglet) [reply src/Controller/CRUDController.php]: I'm gonna merge like this, since I got two approval and don't see a reason to change 3.107 to 3.x to then rechange to 3.107 again if it's ok for you

Comment (phansys) [reply src/Controller/CRUDController.php]: Yes, it isn't a blocker IMO.

**Review by phansys — APPROVED** [review]

**Pull request merged**
